### PR TITLE
add subscription/subscription_item and update spec for card and plan api methods. fix format-encoding for embedded maps in vectors.

### PR DESCRIPTION
**Pull request by humbamp123**

### CODE REVIEW

File: src/clj/stripe/card.clj [hunk 1]
```diff
@@ -110,8 +110,8 @@
       (base-card)
       (ss/stripe-object "card")))
 
-(defn card? [x]
-  (s/valid? ::card x))
+(def card?
+  (partial s/valid? ::car))
```
Comment (jalehman) [inline src/clj/stripe/card.clj]: typo: `::car`

File: src/clj/stripe/card.clj [hunk 2]
```diff
@@ -163,9 +163,11 @@
                  (update opts :params merge params)))))
 
 (s/fdef create!
-        :args (s/cat :customer-id ::id
-                     :source ::source
-                     :opts (s/? h/request-options?))
+        :args (s/alt :binary (s/cat :customer-id ::id
```
Comment (jalehman) [inline src/clj/stripe/card.clj]: `s/alt` isn't necessary in this case. While this is valid, the use of `s/?` is more appropriate.

`s/alt` is needed only when we have more than one optional argument. In this case there's only one.

File: src/clj/stripe/card.clj [hunk 3]
```diff
@@ -177,9 +179,11 @@
    (h/get-req (format "customers/%s/sources/%s" customer-id source-id) opts)))
 
 (s/fdef fetch
-        :args (s/cat :customer-id ::id
-                     :source-id ::id
-                     :opts (s/? h/request-options?))
+        :args (s/alt :binary (s/cat :customer-id ::id
```
Comment (jalehman) [inline src/clj/stripe/card.clj]: see above

File: src/clj/stripe/card.clj [hunk 4]
```diff
@@ -192,10 +196,12 @@
                (assoc opts :params params))))
 
 (s/fdef update!
-        :args (s/cat :customer-id ::id
-                     :card-id ::id
-                     :params ::update-params
-                     :opts (s/? h/request-options?))
+        :args (s/alt :binary (s/cat :customer-id ::id
```
Comment (jalehman) [inline src/clj/stripe/card.clj]: see above

File: src/clj/stripe/card.clj [hunk 5]
```diff
@@ -207,9 +213,11 @@
    (h/delete-req (format "customers/%s/sources/%s" customer-id card-id) opts)))
 
 (s/fdef delete!
-        :args (s/cat :customer-id ::id
-                     :card-id ::id
-                     :opts (s/? h/request-options?))
+        :args (s/alt :binary (s/cat :customer-id :id
```
Comment (jalehman) [inline src/clj/stripe/card.clj]: see above

File: src/clj/stripe/plan.clj [hunk 6]
```diff
@@ -112,8 +123,9 @@
    (h/get-req (str "plans/" (stripe.util.codec/form-encode plan-id)) opts)))
 
 (s/fdef fetch
-        :args (s/cat :plan-id ::id
-                     :opts (s/? h/request-options?))
+        :args (s/alt :unary (s/cat :plan-id ::id)
```
Comment (jalehman) [inline src/clj/stripe/plan.clj]: use `s/?`

File: src/clj/stripe/plan.clj [hunk 7]
```diff
@@ -126,9 +138,11 @@
                (assoc opts :params params))))
 
 (s/fdef update!
-        :args (s/cat :plan-id ::id
-                     :params ::update-params
-                     :opts (s/? h/request-options?))
+        :args (s/alt :binary (s/cat :plan-id ::id
```
Comment (jalehman) [inline src/clj/stripe/plan.clj]: use `s/?`

File: src/clj/stripe/plan.clj [hunk 8]
```diff
@@ -140,8 +154,9 @@
    (h/delete-req (str "plans/" (stripe.util.codec/form-encode plan-id)) opts)))
 
 (s/fdef delete!
-        :args (s/cat :plan-id ::id
-                     :opts (s/? h/request-options?))
+        :args (s/alt :unary (s/cat :plan-id ::id)
```
Comment (jalehman) [inline src/clj/stripe/plan.clj]: use `s/?`

File: src/clj/stripe/subscription.clj [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+
+(defn fetch
+  "Fetch a subscription."
+  ([subscription-id]
+   (fetch subscription-id {}))
+  ([subscription-id opts]
+   (h/get-req (str "subscriptions/" subscription-id))))
+
+(s/fdef fetch
+        :args (s/alt :unary (s/cat :subscription-id ::id)
```
Comment (jalehman) [inline src/clj/stripe/subscription.clj]: `s/?`

File: src/clj/stripe/subscription.clj [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+   (update! subscription-id params {}))
+  ([subscription-id params opts]
+   (h/post-req (str "subscriptions/" subscription-id)
+               (assoc opts :params params))))
+
+(s/fdef update!
+        :args (s/alt :unary (s/cat :subscription-id ::id)
+                     :binary (s/cat :subscription-id ::id
+                                    :params ::update-params)
+                     :ternary(s/cat :subscription-id ::id
```
Comment (jalehman) [inline src/clj/stripe/subscription.clj]: insert a space between `y(`

File: src/clj/stripe/subscription_item.clj [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  plan/plan?)
+
+(s/def ::quantity
+  pos-int?)
+
+(s/def ::subscription
+  string?)
+
+(s/def ::subscription-item
+  (-> (s/keys :req-un [::id ::created ::plan ::quantity
```
Comment (jalehman) [inline src/clj/stripe/subscription_item.clj]: condense to one line--not overflowing 80 chars

File: src/clj/stripe/util/codec.clj [hunk 12]
```diff
@@ -51,7 +67,7 @@
            (mapcat
             (fn [[k v]]
               (cond
-                (or (seq? v) (sequential? v)) (map #(encode-array-param [k %]) v)
+                (or (seq? v) (sequential? v)) (map-indexed #(encode-array-param %1 k %2) v)
```
Comment (jalehman) [inline src/clj/stripe/util/codec.clj]: Is it necessary to use both `seq?` and `sequential?` here? If so, why?

File: src/clj/stripe/plan.clj [hunk 13]
```diff
@@ -126,9 +138,11 @@
                (assoc opts :params params))))
```
Comment (tdiede) [inline src/clj/stripe/plan.clj]: Are params required for update! to plans in the Stripe API?

Comment (humbamp123) [reply src/clj/stripe/plan.clj]: Params are required to have any affect on a plan.

File: src/clj/stripe/plan.clj [hunk 14]
```diff
@@ -112,8 +123,9 @@
    (h/get-req (str "plans/" (stripe.util.codec/form-encode plan-id)) opts)))
```
[the remark this answers is not in the document]
Comment (jalehman) [reply src/clj/stripe/plan.clj]: Actually, it’s probably better to not form encode the plan-id at all. Let’s remove this @humbamp123

**Pull request merged**